core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalReadConverterContext.scala

```diff
@@ -17,59 +17,127 @@
+  private def shortAt(chunk: Chunk) = longAt(chunk).toShort
+  private def floatAt(chunk: Chunk) = longAt(chunk).toFloat
+
+  private def categoricalString(chunk: Chunk) = chunk.vec().domain()(longAt(chunk).toInt)
+
+  private def uuidString(chunk: Chunk) = new java.util.UUID(chunk.at16h(rowIdx), chunk.at16l(rowIdx)).toString
+
+  // TODO(vlad): take care of this bad typing
+  private def timestamp(chunk: Chunk) = longAt(chunk) * 1000L
+
+  private def plainString(chunk: Chunk) = chunk.atStr(new BufferedString(), rowIdx).toString
+
+  val StringProviders = Map[Byte, (Chunk => String)](
+    Vec.T_CAT -> categoricalString,
+    Vec.T_UUID -> uuidString,
+    Vec.T_STR -> plainString
+  ) withDefault((t:Byte) => {
+    assert(assertion = false, s"Should never be here, type is $t")
+    (_:Chunk) => null
+  }
+    )
+
+  def string(chunk: Chunk): String = StringProviders(chunk.vec().get_type())(chunk)
+
+  // TODO(vlad): check if instead of stringification, we could use bytes
+  private def utfString(chunk: Chunk) = UTF8String.fromString(string(chunk))
 
   private def underlyingFrame = DKV.get(Key.make(keyName)).get.asInstanceOf[Frame]
 
-  override def numRows: Int = nrows
+  private val DefaultReader: Reader = (_:Int) => None
+
+  import SupportedTypes._
+  /**
+    * Given a a column number, returns an Option[T]
+    * with the value parsed according to the type.
+    * You can override it.
+    *
+    * A map from type name to option reader
+    */
+  private lazy val ExtractorsTable: Map[SimpleType[_], Chunk => _] = Map(
+    Boolean    -> booleanAt _,
+    Byte       -> byteAt _,
+    Double     -> doubleAt _,
+    Float      -> floatAt _,
+    Integer    -> intAt _,
+    Long       -> longAt _,
+    Short      -> shortAt _,
+    String     -> plainString _,
+    Category   -> categoricalString _,
+    UUID       -> uuidString _,
+    UTF8       -> utfString _,
+    Timestamp  -> timestamp _
+  )
+
+  private lazy val OptionReadersMap: Map[OptionalType[_], OptionReader] =
+    ExtractorsTable map {
+      case (t, getter) => SupportedTypes.byBaseType(t) -> returnOption(getter) _
+    } toMap
+
+  private lazy val OptionReaders: Map[OptionalType[_], OptionReader] = OptionReadersMap withDefault
+    (t => throw new scala.IllegalArgumentException(s"Type $t conversion is not supported in Sparkling Water"))
+
+  private lazy val SimpleReadersMap: Map[SimpleType[_], Reader] =
+    ExtractorsTable map {
+      case (t, getter) => t -> returnSimple(t.defaultValue, getter) _
+    } toMap
+
+  private lazy val SimpleReaders: Map[SimpleType[_], Reader] = SimpleReadersMap withDefault
+    (t => throw new scala.IllegalArgumentException(s"Type $t conversion is not supported in Sparkling Water"))
+
+  private def plainReaderFor(supportedType: SimpleType[_], optReader: OptionReader): Reader =
+    (col: Int) => optReader(col).getOrElse(supportedType.defaultValue)
+
+  lazy val readerMapByName: Map[NameOfType, Reader] = (OptionReaders ++ SimpleReaders) map {
```

Comment:
See, `def` in a trait without implementation is just a general idea of something being provided in implementations. The implementation may be def, or val, or lazy val. Def will reevaluate on every reference, which makes no sense. So either val or lazy val. I made implementation lazy val for the reason of avoiding bad initialization order. Scala people blame JVM, but I would blame scala compiler. It happens pretty frequently that a member of a class refers another member, and if that member is not initialized yet (just because it's below, textually), it is only discovered in runtime, and it is null. Not sure if there is a bug filed, but well, our situation is special, we are targeting a range of compilers, as I understand.